cpp/src/arrow/csv/converter.cc

```diff
@@ -346,6 +347,65 @@ class TimestampConverter : public ConcreteConverter {
   }
 };
 
+/////////////////////////////////////////////////////////////////////////
+// Concrete Converter for Decimals
+
+class DecimalConverter : public ConcreteConverter {
+ public:
+  using ConcreteConverter::ConcreteConverter;
+
+  Status Convert(const BlockParser& parser, int32_t col_index,
+                 std::shared_ptr<Array>* out) override {
+    Decimal128Builder builder(type_, pool_);
+
+    auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+      if (IsNull(data, size, quoted)) {
+        builder.UnsafeAppendNull();
+        return Status::OK();
+      }
+      // Skip trailing whitespace
```

Comment:
Indeed, it might be nice to try and factor this out.

Reply:
Done.